### ISSUE TITLE
Version annotation not adding route condition for API versioning

### ISSUE THREAD

**Issue by angrygreenfrogs** [FriendsOfSymfony/FOSRestBundle#1491]
I'm having trouble with getting the @Version annotation to work. It does not seem to be causing the expected route condition to be added.

I've installed the latest bundle 2.0 (commit f4ce3e9ed88b41c449aa45b120d880723e5f75d7)

I'm trying to make use of the API versioning system with the query parameter style "?version=v1".

I have two separate controllers setup for testing, one for v1 and the other for v2.

I'm trying to make use of the Version annotation to select the correct controller.

config.yml
```
fos_rest:
    versioning:
        enabled: true
        default_version: v1
        resolvers:
            query: 
                enabled: true
                parameter_name: version
    param_fetcher_listener: true
    format_listener:
        rules:
            - { path: '^/api/', fallback_format: json }
            - { path: '^/', stop: true }
```

Version 2 controller
```
/**
 * @Version("v2")
 */
class Testv2Controller extends FOSRestController implements ApiKeyAuthenticatedController
{
    /**
     * Test
     * @Rest\View()
     * @Route("/api/function/test.{_format}",defaults={"_format": "json"})
     * @Method("POST")
     * @RequestParam(name="apikey", requirements="[a-zA-Z0-9]{32}", allowBlank=false, strict=true, description="Application API Key")
     * @param ParamFetcher $paramFetcher
     * @return Response
     * @throws HttpException on application failure
     */
    public function getTestAction(ParamFetcher $paramFetcher)
    {
....
```

Version 1 controller
```
/**
 * @Version("v1")
 */
class Testv1Controller extends FOSRestController implements ApiKeyAuthenticatedController
{
    /**
     * Test
     * @Rest\View()
     * @Route("/api/function/test.{_format}",defaults={"_format": "json"})
     * @Method("POST")
     * @RequestParam(name="apikey", requirements="[a-zA-Z0-9]{32}", allowBlank=false, strict=true, description="Application API Key")
     * @param ParamFetcher $paramFetcher
     * @return Response
     * @throws HttpException on application failure
     */
    public function getTestAction(ParamFetcher $paramFetcher)
    {
....
```

Basically the problem is that calling my API with ?version=v1 or ?version=v2 has no effect and the version 1 controller is always called.

However, if I manually change the routing annotations like this:
v2 route
```
* @Route("/api/function/test.{_format}",defaults={"_format": "json"},condition="request.attributes.get('version') == 'v2'")
```

v1 route
```
* @Route("/api/function/test.{_format}",defaults={"_format": "json"},condition="request.attributes.get('version') == 'v1'")
```

Then it works exactly as expected and the correct function is called. So I'm pretty convinced I've narrowed this down to the @Version annotation not working for me, but for the life of me I can't figure out why.

**Comment by Ener-Getick** [comment]
Is your controller loaded with FOSRestBundle ? (``resource: rest``)

**Comment by angrygreenfrogs** [comment]
Definitely. We're using the rest bundle for just about everything for input, output, etc.

Everything works fine except for the Version annotation not creating the routing configuration

**Comment by Ener-Getick** [comment]
I forgot to backport #1438 in the 2.0 branch...
#1493 should fix this, thanks for the report !

**Comment by angrygreenfrogs** [comment]
Hi guys,

I've checked out a copy of the CHPICK branch where I see these changes that @Ener-Getick  mentioned have been applied. I double-checked to make sure the version I have is consistent and has them in place, which it does.

However, I still can't get this version annotation to work.

I've cut down my test classes to the bare minimum to try to figure it out, but no matter what, it always just uses the 'v1' class.

```
<?php
namespace ApiBundle\Controller\v1;

use FOS\RestBundle\Controller\Annotations\Version;
use FOS\RestBundle\Controller\FOSRestController;
use FOS\RestBundle\Controller\Annotations\Route;
use FOS\RestBundle\Controller\Annotations\View;
use FOS\RestBundle\Controller\Annotations\Post;

/**
 * @Version("v1")
 */
class TestController extends FOSRestController
{
    /**
     * @View
     * @Route("/api/test")
     */
    public function getTestAction()
    {
        $output = array();
        echo "test1";
        $view = $this->view($output, 200);
        return $this->handleView($view);
    }
    
}
```

```
<?php
namespace ApiBundle\Controller\v2;

use FOS\RestBundle\Controller\Annotations\Version;
use FOS\RestBundle\Controller\FOSRestController;
use FOS\RestBundle\Controller\Annotations\Route;
use FOS\RestBundle\Controller\Annotations\View;
use FOS\RestBundle\Controller\Annotations\Post;

/**
 * @Version("v2")
 */
class TestController extends FOSRestController
{
    /**
     * @View
     * @Route("/api/test")
     */
    public function getTestAction()
    {
        $output = array();
        echo "test2";
        $view = $this->view($output, 200);
        return $this->handleView($view);
    }
}
```

I'm sending a request to "/api/test?version=v2", which always outputs "test1", which you can see is from my version 1 class.

I've tried using @Get or @Post instead of @Route, and tried removing @View, but there's no change.

Am I doing something wrong? Any ideas?

**Comment by xabbuh** [comment]
@angrygreenfrogs In your last example the generated route names will likely be the same thus the routes generated from both classes overlap. You may want to use the `@NamePrefix` annotation to overcome this.

**Comment by angrygreenfrogs** [comment]
@xabbuh Thanks for the idea, I've given that a try, but to no effect.

However, again if I manually add the routing condition, it works fine.

So for example this works exactly as expected:

```
<?php
namespace ApiBundle\Controller\v1;

use FOS\RestBundle\Controller\Annotations\Version;
use FOS\RestBundle\Controller\FOSRestController;
use FOS\RestBundle\Controller\Annotations\Route;
use FOS\RestBundle\Controller\Annotations\View;
use FOS\RestBundle\Controller\Annotations\Get;
use FOS\RestBundle\Controller\Annotations\NamePrefix;

/**
 * @Version("v1")
 * @NamePrefix("v1")
 */
class ShareController extends FOSRestController
{
    /**
     * @Route("/api/test",condition="request.attributes.get('version') == 'v1'")
     */
    public function getTestAction()
    {
        $output = array();
        echo "test1";
        $view = $this->view($output, 200);
        return $this->handleView($view);
    }
    
}
```

```
<?php
namespace ApiBundle\Controller\v2;

use FOS\RestBundle\Controller\Annotations\Version;
use FOS\RestBundle\Controller\FOSRestController;
use FOS\RestBundle\Controller\Annotations\Route;
use FOS\RestBundle\Controller\Annotations\View;
use FOS\RestBundle\Controller\Annotations\Get;
use FOS\RestBundle\Controller\Annotations\NamePrefix;

/**
 * @Version("v2")
 * @NamePrefix("v2")
 */
class ShareController extends FOSRestController
{
    /**
     * @Route("/api/test",condition="request.attributes.get('version') == 'v2'")
     */
    public function getTestAction()
    {
        $output = array();
        echo "test2";
        $view = $this->view($output, 200);
        return $this->handleView($view);
    }
}
```

Calling /api/test?version=v1 outputs "test1" and /api/test?version=v2 outputs "test2"

But of course the whole point is I was hoping for the @Version annotation to take care of adding that routing condition for me.

I'm going to try digging through the RestActionReader code to see if I can figure out why the condition isn't being added, but if anyone has any hints about how to debug that, please let me know.

**Comment by angrygreenfrogs** [comment]
Ah, OK, I think I finally figured this out.

The problem is that @Version is incompatible with manually setting your own routes.

As soon as I removed the `* @Route("/api/test")` that I had set, it works just fine as expected.

@Version was creating a separate duplicate route, but it wasn't updating the one that I had manually created.

I'm not quite sure if that's expected functionality, a bug, or a new feature request to add support for manually created routes?

Might be good to note in the documentation for versioning, however.

Also be good to specifically mention that you must have "type: rest" set in the routing configuration.

**Comment by Ener-Getick** [comment]
@angrygreenfrogs This should work for custom routes as long as you only use the ``type: rest`` for your controller.
